lib/identity_cache/memoized_cache_proxy.rb

```diff
@@ -37,101 +38,152 @@ def with_memoization(&block)
     end
 
     def write(key, value)
-      memoized_key_values[key] = value if memoizing?
-      @cache_fetcher.write(key, value)
+      memoizing = memoizing?
+      ActiveSupport::Notifications.instrument('identity_cache.cache.write', memoizing: memoizing) do
+        memoized_key_values[key] = value if memoizing
+        @cache_fetcher.write(key, value)
+      end
     end
 
     def delete(key)
-      memoized_key_values.delete(key) if memoizing?
-      result = @cache_fetcher.delete(key)
-      IdentityCache.logger.debug { "[IdentityCache] delete #{ result ? 'recorded'  : 'failed'  } for #{key}" }
-      result
+      memoizing = memoizing?
+      ActiveSupport::Notifications.instrument('identity_cache.cache.delete', memoizing: memoizing) do
+        memoized_key_values.delete(key) if memoizing
+        result = @cache_fetcher.delete(key)
+        IdentityCache.logger.debug {"[IdentityCache] delete #{ result ? 'recorded' : 'failed'  } for #{key}"}
+        result
+      end
     end
 
     def fetch(key)
-      used_cache_backend = true
-      missed = false
-      value = if memoizing?
-        used_cache_backend = false
-        memoized_key_values.fetch(key) do
-          used_cache_backend = true
-          memoized_key_values[key] = @cache_fetcher.fetch(key) do
-            missed = true
-            yield
+      memo_misses = 0
+      cache_misses = 0
+
+      value = ActiveSupport::Notifications.instrument('identity_cache.cache.fetch') do |payload|
+        payload[:resolve_miss_time] = 0.0
+
+        value = fetch_memoized(key) do
+          memo_misses = 1
+          @cache_fetcher.fetch(key) do
+            cache_misses = 1
+            instrument_duration(payload, :resolve_miss_time) do
```

Comment:
Notifications can be nested. So:

```
instrument("idc.fetch") do
  #do some things

  instrument("idc.resolve_miss_time") do
    # do some things
  end
end
```

If you want to know the duration of the entire fetch it is possible by reading the `idc.fetch`. If you want to know the duration of the miss it is possible by reading the `"idc.resolve_miss_time"`.

Does not that work for you use case?